mkdir_linux.go

```diff
@@ -0,0 +1,224 @@
+				}
+				// Make sure the path is what we expect.
+				fullPath := root + nextPath
+				if err := checkProcSelfFdPath(fullPath, nextDir); err != nil {
+					return nil, "", fmt.Errorf("walking into .. had unexpected result: %w", err)
+				}
+			}
+			_ = currentDir.Close()
+			currentDir = nextDir
+			currentPath = nextPath
+
+		case errors.Is(err, os.ErrNotExist):
+			// We have hit a final component that doesn't exist, so we have our
+			// partial open result. Note that we have to use the OLD remaining
+			// path, since the lookup failed.
+			return currentDir, oldRemainingPath, nil
+
+		case errors.Is(err, unix.ELOOP), errors.Is(err, unix.ENOTDIR):
+			// (O_PATH|O_NOFOLLOW) and O_DIRECTORY means we cannot tell if we
+			// hit a symlink or some other non-directory component from the
+			// error, but we have to readlink the target anyway so we can check
+			// that way so assume it's a symlink.
+			linkDest, err := readlinkatFile(currentDir, part)
+			if err != nil {
+				if errors.Is(err, unix.EINVAL) {
+					// The part was not a symlink, so assume that it's a
+					// regular file. It is possible for it to be a directory
+					// (if an attacker is swapping a directory and
+					// non-directory at this subpath) but erroring out here is
+					// better anyway.
+					err = fmt.Errorf("path component is invalid: %w", unix.ENOTDIR)
+				}
+				return nil, "", err
+			}
+
+			linksWalked++
+			if linksWalked > maxSymlinkLimit {
+				return nil, "", &os.PathError{Op: "partialOpen", Path: root + "/" + unsafePath, Err: unix.ELOOP}
+			}
+
+			remainingPath = linkDest + "/" + remainingPath
+			// Absolute symlinks reset any work we've already done.
+			if path.IsAbs(linkDest) {
+				nextDir, err := dupFile(rootDir)
+				if err != nil {
+					return nil, "", fmt.Errorf("clone root fd: %w", err)
+				}
+				_ = currentDir.Close()
+				currentDir = nextDir
+				currentPath = "/"
+			}
+
+		default:
+			return nil, "", err
+		}
+	}
+	// All of the components existed!
+	return currentDir, "", nil
+}
+
+// MkdirAll is a race-safe alternative to the Go stdlib's os.MkdirAll function,
+// where the new directory is guaranteed to be within the root directory (if an
+// attacker can move directories from inside the root to outside the root, the
+// created directory tree might be outside of the root but the key constraint
+// is that at no point will we walk outside of the directory tree we are
+// creating).
+//
+// Effectively, MkdirAll(root, unsafePath, mode) is equivalent to
+//
+//	path, _ := securejoin.SecureJoin(root, unsafePath)
+//	err := os.MkdirAll(path, mode)
+//
+// But is much safer. The above implementation is unsafe because if an attacker
+// can modify the filesystem tree between SecureJoin and MkdirAll, it is
+// possible for MkdirAll to resolve unsafe symlink components and create
+// directories outside of the root.
+func MkdirAll(root, unsafePath string, mode os.FileMode) error {
```

Comment:
What do you think of doing something simpler, and maybe slower, for the non openat2 case?

Openat2 is supported since 5.6, so lot of distros already have it.

Something like:
```
createdPath := ""
for path in unsafePath { // unsafePath here can be something like strings.Split(unsafePath, string(filepath.Separator))
  
  //NOTE: not exactly as utils.WithProcFd, as that passes a string, we want the fd. So this modified verson should use fh.Fd() instead.
  utils.WithProcFd(root, createdPath), func (basePath int) {
    // TODO: handle the case path exists
    unix.MkdirAt(basePath, path)
   }
  createdPath = createdPath + path
}
```
In my mind, it seems semantically simpler (and simpler to see it is correct) if we handle the non-openat2() case using securejoin.Join.

I guess such a closure in go should work (or can be adapted to take path as param too).

This might be slower due to opening the proc fd, and maybe you can do symlink attacks that live inside the rootfs (I don't think it is a problem, it is as if that change was there from the beginning).

Too bad that SecureJoin doesn't return the fd. Although that might be another option to explore: a SecureJoin() variant that returns the fd too. In essence it would be like this `utils.WithProcFd()` we use in runc, and we just use `securejoinFd()` and `mkdirat()` in a loop.

What do you think?

Reply:
Aside from being considerably less efficient (doing `MkdirAll` this way would result in O(n^2) lookups at least because we have to look up each component each time we add a component), `SecureJoin` is not a safe API to use even if you verify through `/proc/self/fd` for the reasons I outlined in #12.

Also, the current implementation has the same logic after we get the path we want from either `partialOpen` or `partialOpenat2`. You would need to have two entirely different implementations if you do it that way.

> Too bad that SecureJoin doesn't return the fd. Although that might be another option to explore: a SecureJoin() variant that returns the fd too. 

That is exactly what `partialOpen` does, it's just that rather than erroring out if the path doesn't exist it returns the last path it got to and the remaining components. If you don't return that, you get O(n^2) lookups because you need that information for `MkdirAll`.

`partialOpen` is basically a better `SecureJoin`.

Reply:
I suspect you won't like the new symlink stack I needed to add to `partialOpen` to make the behaviour match `openat2`...

Reply:
Heh. The more the complexity grows, the more I think O(n^2) for old kernels (4+ years) might not be so bad :-D

I wish there was an `mkdirat2()` that does the mkdir and returns an fd to the created entry on success (O_PATH, etc.). That would really simplify things and remove races here.

Reply:
And this big c&p in partialOpen feels a little weird, to be honest. That can be a better secureJoin() (without an fd it is not a great interface IMHO), but it feels like mkdir should re-use that machinery instead of c&p.

Clearly secureJoin can be implemented by using `partialOpen()` and in the future we should probably do that. This might be ok for now.

Reply:
> Heh. The more the complexity grows, the more I think O(n^2) for old kernels (4+ years) might not be so bad :-D

After adding the symlink stack, I don't think you could reasonably emulate the symlink behaviour with `SecureJoin`.

> I wish there was an mkdirat2() that does the mkdir and returns an fd to the created entry on success (O_PATH, etc.). That would really simplify things and remove races here.

I suggested it, Al rejected it.

Now that you mention it, we should probably check the owner and permission bits (and whether the directory is empty, I guess) though. If any of those properties are wrong, then we're in a directory the attacker swapped. If the owner and permission bits are the same (and the directory is empty), then even if the attacker did swap the directory, there is no difference in what they can do (if they have the same permissions as us, they can mess with our directory after we created it anyway).

Reply:
(I'll answer to the rest in more detail later, thanks for proving so many details!)

> However, I didn't do it in this PR because the semantics are noticeably different when dealing with dangling symlinks and non-existent paths, and I am not entirely sure you can get all of the semantics to be completely identical.

You mentioned the dangling symlink behaviour changes several times, also in commits IIRC, but never what is the difference. Can you please explain what is exactly the difference?

That is *very* useful to effectively review the PR (to see the behavior matches, to see if tests are covering all the interesting cases, etc.).

Reply:
So, in a case like:

```
% mkdir -p root/a/b
% ln -s a/b/doesnt-exist/c root/link
```

where `link` points to a path that doesn't exist, it is a dangling symlink.

`SecureJoin("root", "link/foo")` will return `a/b/doesnt-exist/c/foo`. This is because all symlinks encountered are expanded and treated as though they were part of the original `unsafePath`. (In the context of `SecureJoin`, this behaviour makes sense -- because it doesn't mind returning non-existent paths, and so a dangling symlink is treated the same as any other broken path.) So in a sense, symlink resolutions are "inatomic".

The old behaviour of `partialOpen` (based on `SecureJoin`) had similar semantics, and would result in `a/b` being opened and then `doesnt-exist/c` being created.

This may seem like totally reasonable behaviour, however it is not possible to easily emulate this with `openat2` because `openat2` has to resolve the path completely, and so a dangling symlink is never resolvable (basically, walking into a symlink is an atomic operation -- either the symlink can be resolved fully or it's dangling and you get `-ENOTDIR`/`-ENOENT`). If we wanted to emulate the `SecureJoin` behaviour with `openat2` we would need to resolve symlinks manually and re-do `openat2` for each component of a symlink (which could lead to a DoS if you had very long paths for each symlink).

`mkdirall: make partialOpen semantically identical to openat2 implementation` is the implementation of this, and the following cases in `testMkdirAllBasic` check this behaviour:

 * All of the `a-fake{1,2,3}` tests.
 * The `link3/deep_dangling{1,2}/foo` tests (which check the case where you have deeper symlink resolution chains).

(The key thing is that the behaviour is identical when using `partialOpen` and when using `openat2`.)

I'll work on some `partialLookupInRoot`-specific tests for this behaviour next week.